### PR TITLE
Port OrbitService to Windows

### PR DESCRIPTION
**Pull request by pierricgimmig**
Make OrbitService, OrbitGrpcServer, ProducerSideChannel and
ProducerSideServer compile on Windows.

Rename ServiceLib -> OrbitServiceLib so that it appears next to
the OrbitService project in VS.

Note that the crash service and the ssh watchdog have not yet
been ported to Windows, but will be in another PR.

Tests: Run OrbitService on Windows and Orbit with "--local".
We can now take a capture and Scheduling and sampling events
are now piped to the client.

### CODE REVIEW

File: src/Service/Windows/OrbitGrpcServer.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Comment (dpallotti) [inline src/Service/Windows/OrbitGrpcServer.cpp:1]:
```suggestion
// Copyright (c) 2021 The Orbit Authors. All rights reserved.
```

File: src/Service/Windows/OrbitGrpcServer.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#include <utility>
+
+#include "CaptureService/CaptureStartStopListener.h"
+#include "WindowsCaptureService/WindowsCaptureService.h"
+#include "WindowsProcessService/ProcessServiceImpl.h"
+#include "grpcpp/ext/proto_server_reflection_plugin.h"
+#include "grpcpp/grpcpp.h"
+#include "grpcpp/health_check_service_interface.h"
+
+ABSL_DECLARE_FLAG(bool, devmode);
```
Comment (dpallotti) [inline src/Service/Windows/OrbitGrpcServer.cpp:22]: What is this for here?

File: src/Service/Linux/OrbitService.cpp [hunk 3]
```diff
@@ -129,19 +132,55 @@ static std::unique_ptr<ProducerSideServer> BuildAndStartProducerSideServer() {
...
+  act.sa_handler = SigintHandler;
+  sigemptyset(&act.sa_mask);
+  act.sa_flags = 0;
+  act.sa_restorer = nullptr;
+  sigaction(SIGINT, &act, nullptr);
+}
+
+}  // namespace
+
+std::string OrbitService::GetLogFilePath() {
```
Comment (dpallotti) [inline src/Service/Linux/OrbitService.cpp:175]: In my opinion this was also clearer in main as it doesn't really have any association with the class. Sure, you avoid an `#ifdef` in `main.cpp`, but `if(WIN32)` is just an `#ifdef` in a different form.

File: src/Service/Windows/OrbitService.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Comment (dpallotti) [inline src/Service/Windows/OrbitService.cpp:1]:
```suggestion
// Copyright (c) 2021 The Orbit Authors. All rights reserved.
```

File: src/Service/Windows/OrbitService.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+BOOL WINAPI CtrlHandler(DWORD event_type) {
+  // Handle the CTRL-C signal.
+  if (event_type == CTRL_C_EVENT || event_type == CTRL_CLOSE_EVENT) {
+    exit_requested = true;
+    return TRUE;
+  }
+
+  // Pass other signals to the next handler.
+  return false;
```
Comment (dpallotti) [inline src/Service/Windows/OrbitService.cpp:39]: You use uppercase `TRUE` but C++ `false`?

File: src/Service/Windows/OrbitService.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  constexpr const char* kUri = "127.0.0.1:1789";
+  std::unique_ptr<ProducerSideServer> producer_side_server = BuildAndStartProducerSideServer(kUri);
+  if (producer_side_server == nullptr) {
+    ERROR("Unable to build and start ProducerSideServer.");
+    return -1;
+  }
+  grpc_server->AddCaptureStartStopListener(producer_side_server.get());
+
+  // Wait for exit_request.
+  while (!exit_requested) {
```
Comment (dpallotti) [inline src/Service/Windows/OrbitService.cpp:101]: This ties the object to a global and a signal handler and makes it impossible to use the class in a different way. `exit_requested` should be passed from the outside, as also mentioned for Linux.

File: src/Service/Windows/OrbitService.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  LOG("**********************************");
+#endif
+
+  std::unique_ptr<OrbitGrpcServer> grpc_server = CreateGrpcServer(grpc_port_, dev_mode_);
+  if (grpc_server == nullptr) {
+    ERROR("Unable to create gRPC server.");
+    return -1;
+  }
+
+  constexpr const char* kUri = "127.0.0.1:1789";
```
Comment (dpallotti) [inline src/Service/Windows/OrbitService.cpp:92]: I think we need a constant for the port.

Comment (dpallotti) [reply src/Service/Windows/OrbitService.cpp:92]: `kProducerUri` or `kProducerSideServerUri`

File: src/Service/Windows/OrbitGrpcServer.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace orbit_service {
+
+using orbit_capture_service::CaptureStartStopListener;
+using orbit_windows_capture_service::WindowsCaptureService;
+using orbit_windows_process_service::ProcessServiceImpl;
+
+namespace {
+
+class OrbitGrpcServerImpl final : public OrbitGrpcServer {
```
Comment (dpallotti) [inline src/Service/Windows/OrbitGrpcServer.cpp:32]: This file is basically the same as the Linux version. It only differs the the members of the class and in the calls to `builder.RegisterService`. Fixes and changes on one side will end up not going to the other side. We need to share code here.

File: src/Service/Windows/OrbitService.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+std::string OrbitService::GetLogFilePath() {
+  std::filesystem::path log_dir = orbit_base::GetExecutablePath() / "logs";
+  std::error_code ec;
+  std::filesystem::create_directory(log_dir, ec);
+  const std::filesystem::path log_file_path = log_dir / "OrbitService.log";
+  return log_file_path.string();
+}
+
+int OrbitService::Run() {
```
Comment (dpallotti) [inline src/Service/Windows/OrbitService.cpp:75]: This file is also very similar to the Linux version. `PrintInstanceVersions` is not here, some addresses are different, and the ssh connection check is removed (why?). Other than that it's basically the same. We need to share code.
One small example: The sleep in the loop was recently changed from one second to 200 milliseconds. Whatever the reason, this is not reflected here. We need this kind of changes to stay in sync.

**Review by dpallotti — COMMENTED** [review]
Too much code duplication in my opinion. Also I don't like how some global logic that belonged in main is now responsibility of an object.

File: src/Service/Windows/OrbitService.cpp (hunk 7 above)
Comment (pierricgimmig) [reply src/Service/Windows/OrbitService.cpp]: Renamed to kProducerSideServerUri

File: src/Service/Windows/OrbitService.cpp (hunk 9 above)
Comment (pierricgimmig) [reply src/Service/Windows/OrbitService.cpp]: The ssh connection hasn't been ported yet. Definitely in the plans, but for another PR.

**Comment by pierricgimmig** [comment]
Hi @dpallotti, thanks for the review. I moved away from splitting the OrbitService into platform-specific files in favor of using ifdefs, which makes it easier to reuse code and is not too messy. Please take another look. Thanks!

**Review by dpallotti — APPROVED** [review]
A couple of comments to address, but looks good in general. I think the `#ifdef`s are totally acceptable.

File: src/Service/main.cpp [hunk 10]
```diff
@@ -39,11 +47,36 @@ void InstallSigintHandler() {
...
+void InstallSigintHandler() {
+  if (!SetConsoleCtrlHandler(CtrlHandler, TRUE)) {
+    ERROR("Calling SetConsoleCtrlHandler: %s", orbit_base::GetLastErrorAsString());
+  }
+}
+
+#endif
+
+std::filesystem::path GetLogFilePath() {
+#ifdef WIN32
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/Service/main.cpp:72]: I agree, fixed.

File: src/Service/OrbitService.cpp [hunk 11]
```diff
@@ -129,20 +151,31 @@ static std::unique_ptr<ProducerSideServer> BuildAndStartProducerSideServer() {
...
 
-  auto producer_side_server = std::make_unique<ProducerSideServer>();
-  LOG("Starting producer-side server at %s",
-      orbit_producer_side_channel::kProducerSideUnixDomainSocketPath);
-  if (!producer_side_server->BuildAndStart(
-          orbit_producer_side_channel::kProducerSideUnixDomainSocketPath)) {
-    ERROR("Unable to start producer-side server");
+  std::string unix_socket_path(orbit_producer_side_channel::kProducerSideUnixDomainSocketPath);
+
+  LOG("Starting producer-side server at %s", unix_socket_path);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/Service/OrbitService.cpp:156]: Yep, fixed.

File: src/Service/OrbitService.cpp [hunk 12]
```diff
@@ -117,7 +122,24 @@ static std::unique_ptr<OrbitGrpcServer> CreateGrpcServer(uint16_t grpc_port, boo
   return grpc_server;
 }
 
-static std::unique_ptr<ProducerSideServer> BuildAndStartProducerSideServer() {
+std::unique_ptr<ProducerSideServer> BuildAndStartProducerSideServer(std::string uri) {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/Service/OrbitService.cpp:125]: done.

File: src/Service/OrbitService.cpp [hunk 13]
```diff
@@ -163,13 +196,16 @@ int OrbitService::Run(std::atomic<bool>* exit_requested) {
...
+#ifdef __linux
   // Make stdin non-blocking.
   fcntl(STDIN_FILENO, F_SETFL, O_NONBLOCK);
+#endif
 
   int return_code = 0;
 
   // Wait for exit_request or for the watchdog to expire.
   while (!(*exit_requested)) {
+#ifdef __linux
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/Service/OrbitService.cpp:208]: Added todo.

**Pull request merged**